File: app/src/main/java/euphoria/psycho/library/DataProvider.java
package euphoria.psycho.library;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.os.Environment;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class DataProvider extends SQLiteOpenHelper {

    private static DataProvider sDataProvider;

    public void addFromClipboard(String tag, String context) {
        Cursor cursor = getReadableDatabase().query("document", new String[]{"count"}, "tag=?", new String[]{tag}, null, null, "count DESC");
        int count = 1;
        if (cursor.moveToNext()) {
            count = cursor.getInt(0) + 1;
        }
        cursor.close();
        ContentValues contentValues = new ContentValues();
        contentValues.put("tag", tag);
        contentValues.put("count", count);
        contentValues.put("content", context);
        getWritableDatabase().insert("document", null, contentValues);
    }

    public void deleteByTag(String tag) {
        //DELETE FROM COMPANY WHERE ID = 7;
        getWritableDatabase().delete("document", "tag=?", new String[]{tag});

    }

    public String exportDocument(String tag) {

        Cursor cursor = getReadableDatabase().rawQuery("SELECT content FROM document WHERE tag = ? ORDER BY count ", new String[]{tag});
        StringBuilder builder = new StringBuilder();
        while (cursor.moveToNext()) {

            // ,count    int i=cursor.getInt(1);
            builder.append(cursor.getString(0)).append("\n\n");
        }

        cursor.close();
        return builder.toString();
    }

    public static DataProvider getInstance(Context context) {
        if (sDataProvider == null) {
            sDataProvider = new DataProvider(context);
        }
        return sDataProvider;
    }

    public static DataProvider getInstance() {

        return sDataProvider;
    }

    public void importDocument(File file) {
        try {

            int threshold = 8000;
            FileInputStream in = new FileInputStream(file);

            InputStreamReader reader = new InputStreamReader(in, Charset.forName("utf8"));
            BufferedReader bufferedReader = new BufferedReader(reader);

            StringBuilder sb = new StringBuilder(threshold);
            String l;
            String tag = FileUtils.getFileNameWithoutExtension(file.getName());
            int count = 0;
            while ((l = bufferedReader.readLine()) != null) {
                if (l.trim().length() > 0)
                    sb.append(l).append("\n\n");
                if (sb.length() > threshold) {

                    insert(tag, sb.toString(), ++count);
                    sb = new StringBuilder(threshold);
                }
            }

            if (sb.length() > 0) {
                insert(tag, sb.toString(), ++count);
            }
            FileUtils.closeSilently(reader);
            FileUtils.closeSilently(bufferedReader);
            FileUtils.closeSilently(in);
//            File dir = new File(Files.getExternalStorageDirectoryPath(".readings"), ".imported");
//            dir.mkdirs();
//            File targetFile = new File(dir, file.getName());
//            if (!targetFile.isFile())
//                file.renameTo(targetFile);
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    public void insert(String tag, String content, int count) {
        ContentValues contentValues = new ContentValues();
        contentValues.put("tag", tag);
        contentValues.put("content", content);
        contentValues.put("count", count);
        getWritableDatabase().insert("document", null, contentValues);
    }

    public void insertArticle(String context) {

        Cursor cursor = getReadableDatabase().query("document", new String[]{"count"}, "tag=?", new String[]{"Stories From Clipboard"}, null, null, "count DESC");
        int count = 1;
        if (cursor.moveToNext()) {
            count = cursor.getInt(0) + 1;
        }
        cursor.close();
        ContentValues contentValues = new ContentValues();
        contentValues.put("tag", "Stories From Clipboard");
        contentValues.put("count", count);
        contentValues.put("content", context);
        getWritableDatabase().insert("document", null, contentValues);
    }

    public int queryCount(String t) {
        Cursor cursor = getReadableDatabase().query("document", new String[]{"count"}, "tag=?", new String[]{t}, null, null, "count DESC");
        int count = 0;
        if (cursor.moveToNext()) {
            count = cursor.getInt(0);
        }
        return count;
    }

    public List<String> listTag() {

        Cursor cursor = getReadableDatabase().rawQuery("SELECT DISTINCT tag FROM document ORDER BY tag", null);

        List<String> list = new ArrayList<>();

        while (cursor.moveToNext()) {
            list.add(cursor.getString(0));
        }
        cursor.close();
        return list;
    }

    @Override
    public void onCreate(SQLiteDatabase sqLiteDatabase) {
        sqLiteDatabase.execSQL("CREATE TABLE IF NOT EXISTS document (tag TEXT ,content TEXT,count INTEGER)");
        sqLiteDatabase.execSQL("CREATE UNIQUE INDEX IF NOT EXISTS `idx_tag_count` ON `document` (`tag` ,`count` ASC)");
        sqLiteDatabase.execSQL("CREATE TABLE IF NOT EXISTS settings (tag TEXT ,count INTEGER,scrollY INTEGER)");
        sqLiteDatabase.execSQL("CREATE UNIQUE INDEX IF NOT EXISTS `idx_tag` ON `settings` (`tag` )");

    }

    @Override
    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {

    }

    public String queryContent(String tag, int count) {
        Cursor cursor = getReadableDatabase().rawQuery("SELECT content FROM document WHERE tag = ? AND count = ?", new String[]{tag, Integer.toString(count)});

        String result = "";

        if (cursor.moveToNext()) {
            result = cursor.getString(0);
        }
        cursor.close();
        return result;
    }

    public List<Integer> queryMatchesContent(String tag, Pattern pattern) {
        Cursor cursor = getReadableDatabase().rawQuery("SELECT count,content FROM document WHERE tag = ?", new String[]{tag});
        List<Integer> list = new ArrayList<>();


        while (cursor.moveToNext()) {
            if (pattern.matcher(cursor.getString(1)).find()) {
                list.add(cursor.getInt(0));
            }
        }
        cursor.close();
        if (list.size() < 1) return null;
        return list;
    }

    public int[] querySettings(String tag) {
        int[] settings = new int[]{
                1, 0
        };
        Cursor cursor = getReadableDatabase().rawQuery("SELECT count,scrollY FROM settings WHERE tag = ?", new String[]{tag});

        if (cursor.moveToNext()) {
            settings[0] = cursor.getInt(0);
            settings[1] = cursor.getInt(1);
        }
        cursor.close();

        return settings;
    }

    public void updateSettings(String tag, int count, int scrollY) {

        ContentValues initialValues = new ContentValues();
        initialValues.put("tag", tag); // the execution is different if _id is 2
        initialValues.put("count", count);
        initialValues.put("scrollY", scrollY);

        int id = (int) getWritableDatabase().insertWithOnConflict("settings", null, initialValues, SQLiteDatabase.CONFLICT_IGNORE);
        if (id == -1) {
            getWritableDatabase().update("settings", initialValues, "tag=?", new String[]{tag});  // number 1 is the _id here, update to variable for your code
        }
    }

    public void updateTag(String tag, String newTag) {

        ContentValues initialValues = new ContentValues();
        initialValues.put("tag", newTag); // the execution is different if _id is 2


        getWritableDatabase().update("document", initialValues, "tag=?", new String[]{tag});  // number 1 is the _id here, update to variable for your code
    }

    public DataProvider(Context context) {

        super(context, new File(new File(Environment.getExternalStorageDirectory(), ".readings"), "datas.db").getAbsolutePath(),
                null, 1);
    }
}

File: app/src/main/java/euphoria/psycho/library/DataProvider.kt
package euphoria.psycho.library

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.os.Environment
import java.io.File

class DataProvider(context: Context = App.instance) : SQLiteOpenHelper(
        context, File(File(Environment.getExternalStorageDirectory(), ".readings"), "datas.db").absolutePath,
        null,
        DataProvider.DATABASE_VERSION
) {
    fun importFile(file: File) {
        val tag = file.nameWithoutExtension
        val list = file.readLines()
        var sb = StringBuilder()
        val length = 8000
        var count = 0;
        for (line in list) {
            if (line.isNotBlank())
                sb.append(line).append('\n')
            if (sb.length > length) {
                insert(tag, sb.toString(), ++count)
                sb = StringBuilder()
            }
        }
        if (sb.isNotEmpty()) {
            insert(tag, sb.toString(), ++count)
        }
    }

    fun insert(tag: String, content: String, count: Int) {
        val v = ContentValues()
        v.put("tag", tag)
        v.put("content", content)
        v.put("count", count)
        writableDatabase.insert("document", null, v)
    }

    fun addFromClipboard(tag: String, context: String) {
        val cursor = readableDatabase.query("document", arrayOf("count"), "tag=?", arrayOf(tag), null, null, "count DESC")
        var count = 1
        if (cursor.moveToNext()) {
            count = cursor.getInt(0) + 1
        }
        cursor.close()
        val contentValues = ContentValues()
        contentValues.put("tag", tag)
        contentValues.put("count", count)
        contentValues.put("content", context)
        writableDatabase.insert("document", null, contentValues)
    }

    fun queryMatchesContent(tag: String, pattern: String): List<Int> {
        val cursor = readableDatabase.rawQuery("SELECT count,content FROM document WHERE tag = ?", arrayOf(tag));
        val list = ArrayList<Int>()
        var regex = Regex(pattern)
        while (cursor.moveToNext()) {
            if (regex.containsMatchIn(cursor.getString(1))) {
                list.add(cursor.getInt(0))
            }
        }
        cursor.close()
        return list
    }

    fun listTag(): ArrayList<String> {
        val cursor = readableDatabase.rawQuery("SELECT DISTINCT tag FROM document ORDER BY tag", null);
        val list = ArrayList<String>();
        while (cursor.moveToNext()) {
            list.add(cursor.getString(0));
        }
        cursor.close();
        return list;
    }

    fun queryCount(tag: String): Int {
        val cursor = readableDatabase.query("document", arrayOf("count"), "tag=?", arrayOf(tag), null, null, "count DESC");
        var count = 0;
        if (cursor.moveToNext()) {
            count = cursor.getInt(0);
        }
        cursor.close()
        return count;
    }

    fun insertArticle(context: String) {
        val cursor = readableDatabase.query("document", arrayOf("count"), "tag=?", arrayOf("Stories From Clipboard"), null, null, "count DESC");
        var count = 1;
        if (cursor.moveToNext()) {
            count = cursor.getInt(0) + 1;
        }
        cursor.close();
        val contentValues = ContentValues();
        contentValues.put("tag", "Stories From Clipboard");
        contentValues.put("count", count);
        contentValues.put("content", context);
        writableDatabase.insert("document", null, contentValues);
    }

    fun queryContent(tag: String, count: Int): String {
        val cursor = readableDatabase.rawQuery("SELECT content FROM document WHERE tag = ? AND count = ?", arrayOf(tag, Integer.toString(count)))
        var result = ""
        if (cursor.moveToNext())
            result = cursor.getString(0)
        cursor.close()
        return result
    }

    fun updateSettings(tag: String, count: Int, scrollY: Int) {
        val v = ContentValues()
        v.put("tag", tag)
        v.put("count", count)
        v.put("scrollY", scrollY)
        writableDatabase.insertWithOnConflict("settings", null, v, SQLiteDatabase.CONFLICT_REPLACE);
    }

    fun querySettings(tag: String): Pair<Int, Int> {
        var count = 1
        var y = 0
        val cursor = readableDatabase.rawQuery("SELECT count,scrollY FROM settings WHERE tag = ?", arrayOf(tag));
        if (cursor.moveToNext()) {
            count = cursor.getInt(0)
            y = cursor.getInt(1)
        }
        cursor.close()
        return count to y
    }

    fun updateDocument(tag: String, count: Int, content: String) {
        val values = ContentValues()

        values.put("content", content)
        writableDatabase.update("document", values, "tag=? and count=?", arrayOf(tag, count.toString()));
    }

    fun updateTag(tag: String, newTag: String) {
        val values = ContentValues()
        values.put("tag", newTag)
        writableDatabase.update("document", values, "tag=?", arrayOf(tag));
    }

    fun deleteByTag(tag: String) {
        writableDatabase.execSQL("VACUUM")
        writableDatabase.delete("document", "tag=?", arrayOf(tag))
    }

    override fun onCreate(sqLiteDatabase: SQLiteDatabase?) {

        sqLiteDatabase?.run {
            execSQL("CREATE TABLE IF NOT EXISTS document (tag TEXT ,content TEXT,count INTEGER)");
            execSQL("CREATE UNIQUE INDEX IF NOT EXISTS `idx_tag_count` ON `document` (`tag` ,`count` ASC)");
            execSQL("CREATE TABLE IF NOT EXISTS settings (tag TEXT ,count INTEGER,scrollY INTEGER)");
            execSQL("CREATE UNIQUE INDEX IF NOT EXISTS `idx_tag` ON `settings` (`tag` )");
        }
    }

    override fun onUpgrade(p0: SQLiteDatabase?, p1: Int, p2: Int) {
    }

    companion object {
        private const val DATABASE_VERSION = 1

        private var instance: DataProvider? = null

        fun getInstance() = instance ?: DataProvider().also { instance = it }


    }
}